app/javascript/onboarding/components/ProfileForm.jsx

```diff
@@ -54,13 +59,125 @@ class ProfileForm extends Component {
     const { name, value } = e.target;
 
     currentFormState[name] = value;
+    this.setState({
+      formValues: currentFormState,
+      canSkip: this.formIsEmpty(currentFormState),
+    });
+  }
+
+  handleColorPickerChange(e) {
+    const { formValues } = { ...this.state };
+    const currentFormState = formValues;
+
+    const field = e.target;
+    const { name, value } = field;
+
+    let sibling = field.nextElementSibling
+      ? field.nextElementSibling
+      : field.previousElementSibling;
+    sibling.value = value;
 
+    currentFormState[name] = value;
+    this.setState({
+      formValues: currentFormState,
+      canSkip: this.formIsEmpty(currentFormState),
+    });
+  }
+
+  formIsEmpty(currentFormState) {
     // Once we've derived the new form values, check if the form is empty
     // and use that value to set the `canSkip` property on the state.
-    const formIsEmpty =
-      Object.values(currentFormState).filter((v) => v.length > 0).length === 0;
+    Object.values(currentFormState).filter((v) => v.length > 0).length === 0;
+  }
 
-    this.setState({ formValues: currentFormState, canSkip: formIsEmpty });
+  checkboxField(field) {
+    return (
+      <div class="crayons-field crayons-field--checkbox">
+        <input
+          class="crayons-checkbox"
+          type="checkbox"
+          name={field.attribute_name}
+          id={field.attribute_name}
+          onChange={this.handleChange}
+        />
+        <label class="crayons-field__label" htmlFor={field.attribute_name}>
+          {field.label}
+        </label>
+        {field.description && (
+          <p class="crayons-field__description">{field.description}</p>
+        )}
+      </div>
+    );
+  }
+
+  textField(field) {
+    return (
+      <div class="crayons-field">
+        <label class="crayons-field__label" htmlFor={field.attribute_name}>
+          {field.label}
+        </label>
+        <input
+          class="crayons-textfield"
+          placeholder={field['placeholder_text']}
+          name={field.attribute_name}
+          id={field.attribute_name}
+          onChange={this.handleChange}
+        />
+        {field.description && (
+          <p class="crayons-field__description">{field.description}</p>
+        )}
+      </div>
+    );
+  }
+
+  textAreaField(field) {
```

Comment:
refactor (blocking): This can be extracted into it's own component. It does not depend on any state of the `<ProfileForm />` component. `field` can be passed in as a prop, e.g.

`<TextField field={field}  />`

Since this is only used in this file, you can keep the component in the same file, or if you want you can put it in its own separate file.

Reply:
@nickytonline I'm happy to do this, but I was curious why you'd put it in its own component (instead of just keeping it in a function) since it's not being reused by any other components.

Reply:
✅